### PR TITLE
[WIP] Fix network hanging issues

### PR DESCRIPTION
**Pull request by sepalani**
This PR should truly fix: https://dolp.in/i11867

What happened (assuming there is no timing issues):
  1. The receive thread is on a blocking `RecvFrom` call
  2. The game calls `SOCleanup` which:
       * fires `IOCTL_NWC24_CLEANUP_SOCKET`
       * calls `IOS_Close` on the network handle
       * sets `soState = 1` (`soState = 2` before this)
  3. The game calls `SOClose` but `ioctlv` isn't fired because `soState = 1`

In other words, `RecvFrom` won't get a response and will wait forever (since Dolphin doesn't receive `IOCTL_SO_CLOSE`), again... 

It might require a hardware test to confirm that's the intended behaviour.

Ready to be reviewed.

### CODE REVIEW

**Comment by sepalani** [comment]
I added a timeout on blocking connect call based on this hardware test. Beware that it tries to connect to an ubisoft server (which timeout after ~3 minutes on hardware).

[wii-connect-timeout.zip](https://github.com/dolphin-emu/dolphin/files/4516755/wii-connect-timeout.zip)

I'll need to reverse IOS to confirm that behaviour, however.

**Comment by sepalani** [comment]
Since all these commits will require appropriate hwtest, I'll split them in different PRs.
